includes/config.php

```diff
@@ -0,0 +1,165 @@
+ * Retrieve the configuration for HM Platform.
+ *
+ * The configuration is defined by merging the defaults with the various files that allow to customise a particular
+ * installation.
+ *
+ * @return array Configuration data.
+ */
+function get_config() {
+	static $config;
+
+	if ( ! $config ) {
+		$config = get_merged_defaults_and_customisations();
+	}
+
+	return $config;
+}
+
+/**
+ * Return the value for a particular setting from the configuration,
+ *
+ * @param string $key Settings key to retrieve the value from.
+ *
+ * @return mixed Settings value.
+ *
+ * @throws Exception if the settings key cannot be found.
+ */
+function get_config_value( string $key ) {
+	$config = get_config();
+
+	if ( ! array_key_exists( $key, $config ) ) {
+		throw new Exception( 'Could not find the ' . $key . ' setting in the configuration.' );
+	}
+
+	return $config[ $key ];
+}
+
+/**
+ * Merge the defaults and the contents of the various configuration files into a single configuration.
+ *
+ * @return array Configuration data.
+ */
+function get_merged_defaults_and_customisations() {
+	$config = get_default_configuration();
+
+	// Look for a `hm` section in `package.json` in the content directory.
+	if ( is_readable( WP_CONTENT_DIR . '/package.json' ) ) {
+		$customisation = get_json_file_contents_as_array( WP_CONTENT_DIR . '/package.json' );
+
+		if ( isset( $customisation['hm'] ) && is_array( $customisation['hm'] ) ) {
+			$config = get_merged_settings( $config, $customisation['hm'] );
+		}
+	}
+
+	// Look for a `hm.json` config file.
+	if ( is_readable( WP_CONTENT_DIR . '/hm.json' ) ) {
+		$config = get_merged_settings( $config, get_json_file_contents_as_array( WP_CONTENT_DIR . '/hm.json' ) );
+	}
+
+	// Look for the environment specific `hm.{env}.json`config file.
+	if ( defined( 'HM_ENV_TYPE' ) && is_readable( WP_CONTENT_DIR . '/hm.' . HM_ENV_TYPE . '.json' ) ) {
+		$config = get_merged_settings( $config, get_json_file_contents_as_array( WP_CONTENT_DIR . '/hm.' . HM_ENV_TYPE . '.json' ) );
+	}
+
+	return $config;
+}
+
+/**
+ * Override settings in an existing configuration file.
+ *
+ * Merge customisations into a configuration file. Existing settings will be overwritten.
+ *
+ * @param array $config    Existing configuration.
+ * @param array $overrides Settings to merge in.
+ *
+ * @return array Configuration data.
+ */
+function get_merged_settings( array $config, array $overrides ) {
+	if ( ! isset( $overrides['plugins'] ) || ! is_array( $overrides['plugins'] ) ) {
+		return $config;
+	}
+
+	$config['plugins'] = get_merged_plugin_settings( $config['plugins'], $overrides['plugins'] );
+
+	return $config;
+}
+
+/**
+ * Merge plugins customisations into a configuration file.
+ *
+ * @param array $config    Existing configuration.
+ * @param array $overrides Settings to merge in.
+ *
+ * @return array Configuration data.
+ */
+function get_merged_plugin_settings( array $config, array $overrides ) {
+	$keys = [ 'appendFile', 'enabled', 'prependFile' ];
```

Comment:
Good call on the white list, though I think `append` and `prepend` should be things we control tightly ourselves for bootstrapping and modifying default plugin behaviour. Users of EK and platform can use the regular method of plugins / themes to add custom behaviour.

Reply:
Makes sense, in that case, I would have these files loaded from the `hm-platform` directory.

Reply:
Not sure if it was implied here, but I think any config should not container the actual PHP file paths, just whether feature 'x' is enabled. The specific plugin file is an implementation detail, not a configuration of hm cloud / EK, as I'd see it.

Reply:
@joehoyle this is the "plugin manifest" so to speak, not the actual config for a project. `hm-platform` needs to know what plugins are available to it and I figured that the manifest would be the logical place to put the info needed to load & bootstrap the plugin. Avoids duplicating the whole manifest in some form in the PHP.

Reply:
It does follow a similar structure to what will be in a project config file to make it easier to set defaults and merge whitelisted overrides.

Reply:
Ahh gotcha, in that case what structure does the hm.json look like for enabling plugins?

Reply:
> It does follow a similar structure to what will be in a project config file to make it easier to set defaults and merge whitelisted overrides.

Ok, what I said might still apply in that case?

Reply:
Yeah perhaps, although platform users wouldn't be able to override the files parts of the manifest. There will be some things to figure out and improve on but I want to start using this asap as it's becoming a blocker.